### DISCUSSION TITLE
[V5]start two different servers (http and https) with the same base and custom ports

### DISCUSSION THREAD

**Discussion by zarbi1** [adonisjs/core#3051] [Help] [answered]
Hello, I would like to start 2 different servers with my same adonisjs code base. For now my server start looks like that:

```
import 'reflect-metadata'
import sourceMapSupport from 'source-map-support'
import { Ignitor } from '@adonisjs/core/build/standalone'
import { createServer } from "https";
import { readFileSync } from 'fs';
import { join } from 'path';

sourceMapSupport.install({ handleUncaughtExceptions: false })

const privateKey = readFileSync(join(__dirname + '/sslCert/server.pem'), 'utf8');
const certificate = readFileSync(join(__dirname + '/sslCert/server.cert'), 'utf8');
new Ignitor(__dirname).httpServer().start((handle) => {
    return createServer(
      {
        key: privateKey,
        cert: certificate,
      },
      handle
    );
  });


```
I would like to start the https server on port 3333 and http server on 4444 if possible (I didn't find a way to start 2 severs for now). The ip is the same, I just want two different way to reach the api

**Comment by feniljariwala82** [comment]
Can you please share your exact problem?

**Reply by zarbi1** [reply]
I would like to start both https and http server with the same program (https server on port 3333 and http server on port 4444)

**Comment by McSneaky** [comment]
Why not start single instance of AdonisJS on port 3333 and then have Nginx or Apache or similar handle HTTPS and HTTP?

This way you have just a single instance and won't run into problems that arise when you start distributing vertically + you have more flexibility

Or if you want to start 2 servers on same IP with different ports, then just update PORT variable:
`PORT=3333 node server.js` and `PORT=4444 node server.js` 
Running those 2 commands will start 2 AdonisJS instances, one on 3333 and other on 4444

**Reply by zarbi1** [reply]
I tried with apache, but when I am trying to interrogate a route with http (example: http://myapi.com/route) I get an `ERR_EMPTY_RESPONSE`
So I think that the server cant support both https and httpon the same port evn with apache (I am not forcing https).

The second solution is the one that I am using but in a different way: I needed http for specific routes, running 2 servers with both https and http routes would be too much, so I created an adonis server for https routes and an other for http routes. For the moment I have no other solution than that. That's why I asked is it was possible to access the server with http and https (so start the server on multiple port without creating an other instance)

**Comment by RomainLanz** [comment]
Hey @zarbi4734! 👋 

The recommended way of handling HTTPS is to use a reverse proxy (like Nginx, Caddy, or Apache) to hold the certificate while the Node.js stays in HTTP.

**Reply by zarbi1** [reply]
Okay but if I do this, can I still send https requests to nodejs ?

**Reply by RomainLanz** [reply]
Why would you want to do that?

If you are using a reverse proxy, you should hit your reverse proxy, not your Node.js application directly.

**Reply by zarbi1** [reply]
I need to have https for my nodejs because my website need to send requests to it, and if it's an http request it wont work for security reasons.

**Comment by McSneaky** [comment]
> So I think that the server cant support both https and httpon the same port evn with apache (I am not forcing https).

Yea, that's true. It's impossible to have HTTPS and HTTP to have on the same ports. By default HTTP uses port 80 and HTTPS uses port 443

_Or well.. It is possible to have both on the same port, but most likely you'll have to write your own TLS handsake and that won't work out-of-box with anything else than your own server and client, won't work in browsers etc_

Btw, out of curiosity. Why you need plain HTTP version in first place?

**Reply by zarbi1** [reply]
Because I am working the the stripe api, and if I want to work with there webhooks I need to receive there request with http because they do not support TLS 1.3, they only support TLS 1.2  or HTTP for there webhooks(which is not very good for the futur because TLS 1.3 is forced with HTTP/3 and all the websites are slowly changing there http version)(my server cant accept tls 1.2 or lower, I tried everything (apache config etc...) but I think it's coming from the certificate provider). 

So for now I am stuck with 2 different servers 😅

I think that I am going to close this discutions because I think that there is no possible solution appart from duplicating the server but that would definitely not be the best optimized solution.

**Comment by zarbi1** [comment, marked as answer]
Closing discussion, the most optimized way to have an http route is to completely create an instance so we dont have to load https routes
